shared/new_sort/quicksort.c

```diff
@@ -0,0 +1,203 @@
+      #define QSWAP     QswapS48
+      #define QUICKSORT QuickSortS48
+      #define QSORT     QSORTS48
+    #endif
+  #else
+    #define INTb      long long 
+    #define SHUFFLEB  Shuffle8
+    #define QSSWAPB   QSswapArr8
+    #if LENC==4
+      #define INTc      int 
+      #define SHUFFLEC  Shuffle4
+      #define QSSWAPC   QSswapArr4
+      #define QSWAP     QswapS84
+      #define QUICKSORT QuickSortS84
+      #define QSORT     QSORTS84
+    #else
+      #define INTc      long long
+      #define SHUFFLEC  Shuffle8
+      #define QSSWAPC   QSswapArr8
+      #define QSWAP     QswapS88
+      #define QUICKSORT QuickSortS88
+      #define QSORT     QSORTS88
+    #endif
+  #endif
+#else
+  #define INTa     long long
+  #define SHUFFLEA Shuffle8
+  #if LENB==4
+    #define INTb      int 
+    #define SHUFFLEB  Shuffle4
+    #define QSSWAPB   QSswapArr4
+    #if LENC==4
+      #define INTc      int
+      #define SHUFFLEC  Shuffle4
+      #define QSSWAPC   QSswapArr4
+      #define QSWAP     QswapL44
+      #define QUICKSORT QuickSortL44
+      #define QSORT     QSORTL44
+    #else
+      #define INTc      long long
+      #define SHUFFLEC  Shuffle8
+      #define QSSWAPC   QSswapArr8
+      #define QSWAP     QswapL48
+      #define QUICKSORT QuickSortL48
+      #define QSORT     QSORTL48
+    #endif
+  #else
+    #define INTb      long long
+    #define SHUFFLEB  Shuffle8
+    #define QSSWAPB   QSswapArr8
+    #if LENC==4
+      #define INTc      int
+      #define SHUFFLEC  Shuffle4
+      #define QSSWAPC   QSswapArr4
+      #define QSWAP     QswapL84
+      #define QUICKSORT QuickSortL84
+      #define QSORT     QSORTL84
+    #else
+      #define INTc      long long
+      #define SHUFFLEC  Shuffle8
+      #define QSSWAPC   QSswapArr8
+      #define QSWAP     QswapL88
+      #define QUICKSORT QuickSortL88
+      #define QSORT     QSORTL88
+    #endif
+  #endif
+#endif
+
+void QSWAP(INTa a[], INTb b[], INTc c[], 
+	   int i, int j, int m, int n, int q)
+{
+   if(i!=j) {
+     INTa s;
+
+     SWAP(a[i],a[j]);
+
+     if(b && n!=0) (void)QSSWAPB(b,i,j,m,n);
+     if(c && q!=0) (void)QSSWAPC(c,i,j,m,q);
+   }
+}
+
+
+
+
+void QUICKSORT(INTa a[], INTb b[], INTc c[], 
+	       int l, int r, int m, int n, int q)
+{
+  int len=r-l+1;
+
+  if (len<=1) {
+    return;
+  } 
+  else if (len==2) {
+    if (a[l]>a[r]) QSWAP(a,b,c,l,r,m,n,q);
+    return;
+  }
+  else {
+    int  j = r;
+    int  i = l-1;
+    INTa v = a[r];
+
+    for(;;)  {
+      while(a[++i]<v && i<r);
+      while(a[--j]>v && j>l);
+      if (j<=i) break;
+      QSWAP(a,b,c,i  ,j,m,n,q);
+    }
+    QSWAP    (a,b,c,i  ,r,m,n,q);
+    QUICKSORT(a,b,c,l  ,j,m,n,q);
+    QUICKSORT(a,b,c,i+1,r,m,n,q);
+  }
+}
+
+
+#define MAX_SORT_LEN 400000000
+
+
+
+// If the length of the sort exceeds a critical value (MAX_SORT_LEN),
+// it halves it repeatedly until it is below this value, does
+// order LEN/MAX_SORT_LEN sorts, and shuffles these in ascending order.
```

Comment:
This comment leaves out the "why".   Is it to give better performance?  To avoid some memory limitation?

Reply:
Only Max could answer "why". My guess is for better performance... I do not think we exceed the threshold currently. Which brings the question, if the "swap-and-shuffle" scheme was well tested

Reply:
I remember one thing. I changed the sort so it pickd the middle item as pivot point. Before that, it chose the first one as the pivot point. So if an array is already sorted, it takes O( n^2). If the array is very large, it would be very slow. It helps if it limits the max size.